File: 240711/트리의 지름/diameter-of-tree.java
import java.util.*;
import java.io.*;

public class Main {
    static int idx, max;
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int N = Integer.parseInt(br.readLine());
        
        ArrayList<int[]>[] tree = new ArrayList[N+1];

        for(int i = 1; i <= N; i++){
            tree[i] = new ArrayList();
        }

        for(int i = 0; i < N-1; i++){
            String[] inputs = br.readLine().split(" ");

            int src = Integer.parseInt(inputs[0]);
            int dst = Integer.parseInt(inputs[1]);
            int val = Integer.parseInt(inputs[2]);

            tree[src].add(new int[]{dst, val});
            tree[dst].add(new int[]{src, val});
        }

        boolean[] v = new boolean[N+1];
        v[2] = true;
        dfs(2, 0, v, tree);
        // System.out.println("new dfs!!!");
        max = 0;
        v = new boolean[N+1];
        v[idx] = true;
        dfs(idx, 0, v, tree);

        System.out.println(max);
    }

    private static void dfs(int cur, int sum, boolean[] v, ArrayList<int[]>[] tree){
        // System.out.printf("current : %d, sum : %d\n", cur, sum);
        for(int[] node : tree[cur]){
            if(!v[node[0]]){
                v[node[0]] = true;
                dfs(node[0], sum+node[1], v, tree);
                v[node[0]] = false;
            }
        }

        // 만약 다 방문했다면, 길이 비교
        if(max < sum){
            max = sum;
            idx = cur;
            // System.out.printf("new max! idx : %d, sum : %d\n", cur, max);
        }
    }
}

File: 240711/트리의 지름/diameter-of-tree.py
n = int(input())
edges = [[] for _ in range(n+1) ]
visited = [0 for _ in range(n+1)]
for _ in range(n-1):
    e1,e2,w = map(int,input().split())
    edges[e1].append((e2,w))
    edges[e2].append((e1,w))


ans =0
idx =0
def dfs(curr,acc,visited,edges):
    global ans,idx
    for node in edges[curr]:
        if not visited[node[0]]:
            visited[node[0]]=1
            dfs(node[0],acc+node[1],visited,edges)
            visited[node[0]]=0
    
    if ans < acc :
        ans =acc
        idx = curr
    return idx, ans
start = dfs(1,0,visited,edges)
visited[1] =1
visited = [0 for _ in range(n+1)]
ans =0
visited[start[0]] =1
end=dfs(start[0],0,visited,edges)
print(end[1])